documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+`Name` is a rough description of the input device. Its value is not intrinsically meaningful.
+
+All devices originate from a context.
+
+```cs
+public interface IInputBackend
+{
+    string Name { get; }
+    nint Handle { get; }
+    IReadOnlyList<IInputDevice> Devices { get; }
+    void Update();
+}
+```
+
+`Name` is a rough description of the input backend. Its value is not intrinsically meaningful.
+
+`Handle` is a globally-unique handle to memory for this input backend. The memory this handle actually points to is implementation-defined.
+
+`Devices` enumerates all of the **connected** devices available from this input backend. When a device is disconnected, its `IInputDevice` object should be discarded by all that consumed it, as it can not be relied upon for being reused by the input backend. An implementation is welcome to reuse old objects, but this is strictly implementation-defined. A device not being present in the `Devices` list is sufficient evidence that a device has been disconnected.
+
+`Update` will update the state of all devices contained within this input backend. The value of the `State` properties on each device must not change until this method is called. This is a departure from 1.0's and 2.0's model of updating state as soon as new information is available, which has resulted in lots of inconsistencies in the past.
+
+All of the `Devices` and `Update`s are aggregated and coordinated by a central input context object.
+
+```cs
+public partial class InputContext
+{
+    public Mice Mice { get; }
+    public Keyboards Keyboards { get; }
+    public Gamepads Gamepads { get; }
+    public Joysticks Joysticks { get; }
+    public IReadOnlyList<IInputDevice> Devices { get; }
+    public List<IInputBackend> Backends { get; }
+    public void Update();
+}
+```
+
+The central input object acts as the main entry point into the Input API, and is responsible for comparing the state reported by the devices for differences between `Update` calls (raising events as necessary). 
+
+`Mice`, `Keyboards`, `Gamepads`, and `Joysticks` are all custom `IReadOnlyList` types for enumerating the **state** (given that the actual input objects are thin, and the state contains a reference to the original device anyway). However, these custom types also contain the events. This is so we can "scope" the events, rather than putting them at the top-level and having to call the events `MouseButtonDown`, `JoystickButtonDown`, etc.
+
+By virtue of the `State` properties not updating until `IInputBackend.Update` is called, the states enumerated by the lists will not change until `Update` is called.
+
+`Update` will call `IInputBackend.Update` on each of the `Backends`, and retrieve the state for each device contained in each backend. As it is doing this, it updates the lists within the input context.
```

Comment:
See the above Update comment, I feel this is even "worse" in terms of threading / safety concerns.

Reply:
Addressed above: the onus is on the user to use this type in a thread safe manor.